File: examples/Google.Cloud.Functions.Examples.SimpleDependencyInjection/Function.cs
// Copyright 2020, Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using Google.Cloud.Functions.Framework;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using System.Threading.Tasks;

namespace Google.Cloud.Functions.Examples.SimpleDependencyInjection
{
    /// <summary>
    /// Simple example of dependency injection, with no additional configuration.
    /// Loggers are provided out-of-the-box by ASP.NET Core, which means you can inject a logger
    /// into your function constructor.
    /// </summary>
    public class Function : IHttpFunction
    {
        private readonly ILogger<Function> _logger;

        public Function(ILogger<Function> logger) =>
            _logger = logger;

        public async Task HandleAsync(HttpContext context)
        {
            _logger.LogInformation("Function called with path {path}", context.Request.Path);
            await context.Response.WriteAsync("Written the request path to the logger provided by dependency injection.");
        }
    }
}

File: src/Google.Cloud.Functions.Templates/templates/http-function-fs/Function.fs
namespace MyFunction

open Google.Cloud.Functions.Framework
open Microsoft.AspNetCore.Http

type Function() =
    interface IHttpFunction with
        /// <summary>
        /// Logic for your function goes here.
        /// </summary>
        /// <param name="context">The HTTP context, containing the request and the response.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        member this.HandleAsync context =
            async {
                do! context.Response.WriteAsync "Hello, Functions Framework." |> Async.AwaitTask
            } |> Async.StartAsTask :> _
